### PR TITLE
Significant overhaul to container+dependency management

### PR DESCRIPTION
**Pull request by jmchilton**
*Overview*

- Migrate "manage dependencies" UI to APIs and VueJS.
  - Add new `summarize_toolbox` API endpoint to mimic information produced by former grid.
  - The tool and requirement views are in the same grid now - depending on what you are viewing there is a "collapse tools" / "expand tools" button to swap to the other view.
  - Unused environments grid migrated as its own component.
  - By default information displayed information is more compact but every row can be expanded and many elements display a lot more information on mouse over.
- Implement fallback container resolvers (describe below).
- Implement RESTful API around container resolvers - describe both how resolvers are configured and the resolution of particular tools (or all tools). Add API endpoint for building containers as well.
- Implement new VueJS table describing container resolution independent of dependencies. Allow mixed mode (show whatever container first resolved) as well as filtering by Docker/Singularity.
  - This information about containers is much more correct than what the manage dependencies grid would display - which was independent of tool IDs and based only on requirements, would not query quay.io for information, etc..
  - Since the grid is restricted to containers, embedded information is richer - displays information about the container resolver, a link to the container, correct table header, shell used by the container, etc..

*Why Fallback Container Resolvers?*

tl;dr - we can write better config files and more useful UIs if we specify default containers this way.

Default containers can be specified in the job_conf but Galaxy doesn't really have the ability to dispatch or reason about such configuration until it is time to run the job. galaxy.tool_util has no clue about job configuration structure and job configuration structure can be (and often is) completely dynamic. As a result, such configuration should be thought of procedurally specified - configuration via container resolvers is effectively more explicitly annotated.

This has two upshots for my current cloud dev push:

- Containers specified this way can reasoned about in terms of the dependency magnagement API so we can provide an interface for seeing what tools will map to job (by default anyway - same stipulation applies to mulled containers which can also be overridden). I will utilize to make #8741.
- While I think job_conf should ultimately allow overriding the container, I think the way this works out in most setups - including the CloudVE setup (https://github.com/galaxyproject/galaxy-helm/blob/master/galaxy/values.yaml#L201) the container is more realistically a global configuration setting. As such it is not naturally specified job_conf, awkardness results, and it would be much more cleanly specified in the container configuration - a global configuration.

*Existing UI in 19.09*

Screen Shot 2019-10-02 at 12 40 18 PM

*New UI*

Screen Shot 2019-11-07 at 3 18 52 PM

---

Screen Shot 2019-11-07 at 3 19 05 PM

---

Screen Shot 2019-11-07 at 3 19 21 PM

---

Screen Shot 2019-11-07 at 3 19 33 PM

---

Screen Shot 2019-11-07 at 3 19 41 PM

---

Screen Shot 2019-11-07 at 3 20 01 PM

---

Screen Shot 2019-11-07 at 3 20 11 PM

---

Screen Shot 2019-11-07 at 3 20 27 PM

---

Screen Shot 2019-11-07 at 3 21 47 PM

*Newer UI (earlier iterations)*

Screen Shot 2019-10-11 at 4 36 58 PM

---

Screen Shot 2019-10-11 at 4 34 39 PM

---

Screen Shot 2019-10-11 at 4 34 25 PM

---

Screen Shot 2019-10-11 at 4 33 47 PM

### CODE REVIEW

**Comment by martenson** [comment]
@jmchilton Could you rebase this please so we get a clean diff?

**Comment by mvdbeek** [comment]
> Default containers can be specified in the job_conf but Galaxy doesn't really have the ability to dispatch or reason about such configuration until it is time to run the job. galaxy.tool_util has no clue about job configuration structure and job configuration structure can be (and often is) completely dynamic. As a result, such configuration should be thought of procedurally specified - configuration via container resolvers is effectively more explicitly annotated.

That is awesome, I agree completely!



> * While I think job_conf should ultimately allow overriding the container, I think the way this works out in most setups - including the CloudVE setup (https://github.com/galaxyproject/galaxy-helm/blob/master/galaxy/values.yaml#L201) the container is more realistically a global configuration setting. As such it is not naturally specified job_conf, awkardness results, and it would be much more cleanly specified in the container configuration - a global configuration.

💯

**Comment by jmchilton** [comment]
@mvdbeek Thanks for the note about container resolvers - glad you agree. I've been holding off a bit a push to implement per tool default containers at the resolver level but I think with you buy in I feel more comfortable adding that.

Have you played with this at all? I feel like there are a lot of next steps here but I think I'd like to undo placing containers and dependencies in the same grid (reasons outlined here https://github.com/galaxyproject/galaxy/pull/7125#issuecomment-539516275) but if you'd like to keep the mixed mode I'll reconsider things.

File: lib/galaxy/webapps/galaxy/api/container_resolution.py [hunk 1]
```diff
@@ -0,0 +1,121 @@
+"""
+API operations allowing clients to manage tool dependencies.
```
Comment (mvdbeek) [inline lib/galaxy/webapps/galaxy/api/container_resolution.py]:
```suggestion
API operations allowing clients to manage container resolution.
```

**Comment by mvdbeek** [comment]
I have now, and this is great! I'm totally fine with separating containers and dependencies, you've convinced me. One thing though is even with the new caching it takes more than an hour to list the container resolver view. Maybe we can just search for containers on demand ?
I checked the quay.io API and didn't really see an alternative for bulk requests (short of just building the list in a cronjob on our side).

Also the resolution view is a lot slower than the legacy view. The legacy view takes 3.5 seconds to load, the new view takes 51 seconds. That should be fixed by https://github.com/jmchilton/galaxy/pull/75

**Comment by mvdbeek** [comment]
Expanding unsresolved tools without requirements also crashes in the JS console:
```
vue.runtime.esm.js?2b0e:619 [Vue warn]: Error in render: "TypeError: Cannot read property 'dependency_resolver' of undefined"

found in

---> <ResolutionDetails> at galaxy/scripts/components/admin/Dependencies/ResolutionDetails.vue
       <BTableCell>
         <BTr>
           <BTbody>
             <BTable>
               <DependencyIndexWrapper> at galaxy/scripts/components/admin/Dependencies/DependencyIndexWrapper.vue
                 <ResolutionIndex> at galaxy/scripts/components/admin/Dependencies/ResolutionIndex.vue
                   <Root>
warn @ vue.runtime.esm.js?2b0e:619
logError @ vue.runtime.esm.js?2b0e:1884
globalHandleError @ vue.runtime.esm.js?2b0e:1879
handleError @ vue.runtime.esm.js?2b0e:1839
Vue._render @ vue.runtime.esm.js?2b0e:3544
updateComponent @ vue.runtime.esm.js?2b0e:4060
get @ vue.runtime.esm.js?2b0e:4473
Watcher @ vue.runtime.esm.js?2b0e:4462
mountComponent @ vue.runtime.esm.js?2b0e:4067
Vue.$mount @ vue.runtime.esm.js?2b0e:8409
init @ vue.runtime.esm.js?2b0e:3118
createComponent @ vue.runtime.esm.js?2b0e:5972
createElm @ vue.runtime.esm.js?2b0e:5919
createChildren @ vue.runtime.esm.js?2b0e:6047
createElm @ vue.runtime.esm.js?2b0e:5948
patch @ vue.runtime.esm.js?2b0e:6471
Vue._update @ vue.runtime.esm.js?2b0e:3939
updateComponent @ vue.runtime.esm.js?2b0e:4060
get @ vue.runtime.esm.js?2b0e:4473
Watcher @ vue.runtime.esm.js?2b0e:4462
mountComponent @ vue.runtime.esm.js?2b0e:4067
Vue.$mount @ vue.runtime.esm.js?2b0e:8409
init @ vue.runtime.esm.js?2b0e:3118
createComponent @ vue.runtime.esm.js?2b0e:5972
createElm @ vue.runtime.esm.js?2b0e:5919
createChildren @ vue.runtime.esm.js?2b0e:6047
createElm @ vue.runtime.esm.js?2b0e:5948
patch @ vue.runtime.esm.js?2b0e:6471
Vue._update @ vue.runtime.esm.js?2b0e:3939
updateComponent @ vue.runtime.esm.js?2b0e:4060
get @ vue.runtime.esm.js?2b0e:4473
Watcher @ vue.runtime.esm.js?2b0e:4462
mountComponent @ vue.runtime.esm.js?2b0e:4067
Vue.$mount @ vue.runtime.esm.js?2b0e:8409
init @ vue.runtime.esm.js?2b0e:3118
createComponent @ vue.runtime.esm.js?2b0e:5972
createElm @ vue.runtime.esm.js?2b0e:5919
updateChildren @ vue.runtime.esm.js?2b0e:6210
patchVnode @ vue.runtime.esm.js?2b0e:6313
patch @ vue.runtime.esm.js?2b0e:6476
Vue._update @ vue.runtime.esm.js?2b0e:3942
updateComponent @ vue.runtime.esm.js?2b0e:4060
get @ vue.runtime.esm.js?2b0e:4473
run @ vue.runtime.esm.js?2b0e:4548
flushSchedulerQueue @ vue.runtime.esm.js?2b0e:4304
eval @ vue.runtime.esm.js?2b0e:1980
flushCallbacks @ vue.runtime.esm.js?2b0e:1906
Promise.then (async)
timerFunc @ vue.runtime.esm.js?2b0e:1933
nextTick @ vue.runtime.esm.js?2b0e:1990
queueWatcher @ vue.runtime.esm.js?2b0e:4396
update @ vue.runtime.esm.js?2b0e:4538
notify @ vue.runtime.esm.js?2b0e:730
reactiveSetter @ vue.runtime.esm.js?2b0e:1055
showRowDetails @ DependencyIndexMixin.js?efce:42
invokeWithErrorHandling @ vue.runtime.esm.js?2b0e:1854
invoker @ vue.runtime.esm.js?2b0e:2179
invokeWithErrorHandling @ vue.runtime.esm.js?2b0e:1854
Vue.$emit @ vue.runtime.esm.js?2b0e:3882
rowClicked @ mixin-tbody-row.js?c324:164
eval @ mixin-tbody-row.js?c324:95
invokeWithErrorHandling @ vue.runtime.esm.js?2b0e:1854
invoker @ vue.runtime.esm.js?2b0e:2179
original._wrapper @ vue.runtime.esm.js?2b0e:6911
Show 17 more frames
vue.runtime.esm.js?2b0e:1888 TypeError: Cannot read property 'dependency_resolver' of undefined
    at Proxy.render (ResolutionDetails.vue?e2f2:71)
    at VueComponent.Vue._render (vue.runtime.esm.js?2b0e:3542)
    at VueComponent.updateComponent (vue.runtime.esm.js?2b0e:4060)
    at Watcher.get (vue.runtime.esm.js?2b0e:4473)
    at new Watcher (vue.runtime.esm.js?2b0e:4462)
    at mountComponent (vue.runtime.esm.js?2b0e:4067)
    at VueComponent.Vue.$mount (vue.runtime.esm.js?2b0e:8409)
    at init (vue.runtime.esm.js?2b0e:3118)
    at createComponent (vue.runtime.esm.js?2b0e:5972)
    at createElm (vue.runtime.esm.js?2b0e:5919)
    at createChildren (vue.runtime.esm.js?2b0e:6047)
    at createElm (vue.runtime.esm.js?2b0e:5948)
    at VueComponent.patch [as __patch__] (vue.runtime.esm.js?2b0e:6471)
    at VueComponent.Vue._update (vue.runtime.esm.js?2b0e:3939)
    at VueComponent.updateComponent (vue.runtime.esm.js?2b0e:4060)
    at Watcher.get (vue.runtime.esm.js?2b0e:4473)
    at new Watcher (vue.runtime.esm.js?2b0e:4462)
    at mountComponent (vue.runtime.esm.js?2b0e:4067)
    at VueComponent.Vue.$mount (vue.runtime.esm.js?2b0e:8409)
    at init (vue.runtime.esm.js?2b0e:3118)
    at createComponent (vue.runtime.esm.js?2b0e:5972)
    at createElm (vue.runtime.esm.js?2b0e:5919)
    at createChildren (vue.runtime.esm.js?2b0e:6047)
    at createElm (vue.runtime.esm.js?2b0e:5948)
    at VueComponent.patch [as __patch__] (vue.runtime.esm.js?2b0e:6471)
    at VueComponent.Vue._update (vue.runtime.esm.js?2b0e:3939)
    at VueComponent.updateComponent (vue.runtime.esm.js?2b0e:4060)
    at Watcher.get (vue.runtime.esm.js?2b0e:4473)
    at new Watcher (vue.runtime.esm.js?2b0e:4462)
    at mountComponent (vue.runtime.esm.js?2b0e:4067)
    at VueComponent.Vue.$mount (vue.runtime.esm.js?2b0e:8409)
    at init (vue.runtime.esm.js?2b0e:3118)
    at createComponent (vue.runtime.esm.js?2b0e:5972)
    at createElm (vue.runtime.esm.js?2b0e:5919)
    at updateChildren (vue.runtime.esm.js?2b0e:6210)
    at patchVnode (vue.runtime.esm.js?2b0e:6313)
    at VueComponent.patch [as __patch__] (vue.runtime.esm.js?2b0e:6476)
    at VueComponent.Vue._update (vue.runtime.esm.js?2b0e:3942)
    at VueComponent.updateComponent (vue.runtime.esm.js?2b0e:4060)
    at Watcher.get (vue.runtime.esm.js?2b0e:4473)
    at Watcher.run (vue.runtime.esm.js?2b0e:4548)
    at flushSchedulerQueue (vue.runtime.esm.js?2b0e:4304)
    at Array.eval (vue.runtime.esm.js?2b0e:1980)
    at flushCallbacks (vue.runtime.esm.js?2b0e:1906)
```

**Comment by jmchilton** [comment]
Thanks for the details review @mvdbeek and thanks for fixing the performance of the resolution grid.

> I'm totally fine with separating containers and dependencies, you've convinced me. 

Let me make that change then and see what this looks like. I'll keep the mixed resolution mode in as an optional API parameter that is marked as experimental and may go away.

>  One thing though is even with the new caching it takes more than an hour to list the container resolver view. Maybe we can just search for containers on demand ?

I have many thoughts on this... my incremental recommendation would be...

- Merge without worrying more about performance initially. The APIs are important and pretty good, we need a way to access this information and this works just fine as is for small toolboxes (Galaxy flavors, targeted public instances, planemo serving a workflow, etc...). It also shouldn't be a problem for Singularity resolution (?) which usegalaxy.* will likely leverage first. Even for Docker - I don't really feel there is another good way to access this information currently - so an hour is better than nothing. The use case I developed this for is checking cloud instances - we should only need to check this information once in a while for QC purposes and such, it isn't an every day thing.
- Iteration 2: Implement tool ID filtering client-side - it is valuable regardless of performance. Doesn't help performance immediately but when paired with iteration 3...
- Iteration 3: Render the grid with the tool ids and bring in the data on demand if there are more than say 500 tools?
   - Iteration 3 - just load the tool IDs without resolution information.
   - Iteration 3.1 - do resolution without the slow resolvers (just ``mulled`` currently), render the information with a spinner next to it implying more complete information is coming.
   - Iteration 3.2 - use client caching to render saved information the same way as 3.1.

  So render what you can, make sure it is marked as incomplete until a full resolution is done allow filtering immediately to prune the space of things that need to be resolved.
   The problem with rendering the resolution only after you ask for it is that I developed this with quickly summarizing a whole toolbox in mind. Maybe it is the right approach though, I will readily admit my proposal above might look like a denial of service against 

- Iteration 4 (though it can be concurrent with client approaches): Cache the ``mulled`` resolution of a tool id on the backend. We should be doing this anyway for reproducibility I think? I would be much less concerned about the hour loading time if it only happened once per instance - we could resolve on install also to further mitigate this.

Typing this all out though I don't know if Iteration 4 shouldn't be Iteration 1. That could be a big win.

This is a fun project and I really love Vue and think our client infrastructure gives us the tools to do this easily... but none of this is on the project roadmap so I'm not sure it is worth the effort immediately.

Update: To be clear here - I mean a semi-permanent cache using Beaker like we do for citations. This is a very similar problem to citations really.

**Comment by jmchilton** [comment]
I've taken multiple passes at improving the UI - the detailed view is a lot more structured now, better presentation, added information about Conda resolver configuration if resolved that way, and added a little button that display underlying JSON for all the objects displayed - that has even more information than presented in the GUI. I think that level of detail is important for developers or admins that this UI is targeted at.

I've removed containers from the "resolution" grid and renamed it "Dependencies". The API still allows mixed resolution if one just adds "include_containers" as a query parameter. We could therefore create a mixed resolution grid again very quickly - either as a new tab or with a checkbox.

I've implemented the Beaker cache for mulled resolution and a per-request miss cache (so null resolutions aren't repeatedly checked).

**Comment by jmchilton** [comment]
I think this ready to go on my end, @mvdbeek anything else you need to see from this PR?

I think this sets up to do some nice fixes as bug fixes during the freeze of the next release if we need to.

File: lib/galaxy/tool_util/deps/containers.py [hunk 2]
```diff
@@ -181,6 +194,19 @@ def __init__(self, app_info):
         self.enable_mulled_containers = app_info.enable_mulled_containers
         self.app_info = app_info
         self.container_resolvers = self.__build_container_resolvers(app_info)
+        self.mulled_resolution_cache = None
+        if app_info.mulled_resolution_cache_type:
+            # Do not want to make beaker a required dependency of container resolution, that doesn't make a lot
```
Comment (mvdbeek) [inline lib/galaxy/tool_util/deps/containers.py]: Would you be opposed to making this a top-level import with try / except ImportError ?

File: lib/galaxy/tool_util/deps/mulled/util.py [hunk 3]
```diff
@@ -47,24 +50,56 @@ def quay_repository(namespace, pkg_name):
...
+    return repo_name in repo_names
+
+
+def mulled_tags_for(namespace, image, tag_prefix=None, resolution_cache=None):
     """Fetch remote tags available for supplied image name.
 
     The result will be sorted so newest tags are first.
     """
+    if resolution_cache is not None:
+        # Following check is pretty expensive against biocontainers... don't even both doing it
```
Comment (mvdbeek) [inline lib/galaxy/tool_util/deps/mulled/util.py]:
```suggestion
        # Following check is pretty expensive against biocontainers... don't even bother doing it
```

**Review by mvdbeek — APPROVED** [review]

File: lib/galaxy/tool_util/deps/containers.py (hunk 2 above)
Comment (jmchilton) [reply lib/galaxy/tool_util/deps/containers.py]: How about https://github.com/galaxyproject/galaxy/pull/8741/commits/b6b67f1f474e9d9cdeb07e06176338d493c86fb0 - cleaner imports, cleaner dependencies?

Comment (mvdbeek) [reply lib/galaxy/tool_util/deps/containers.py]: 😍

**Comment by mvdbeek** [comment]
Arg, there's a conflict here, if you resolve I'll merge

**Comment by jmchilton** [comment]
That inconsiderate jerk @jmchilton is constantly inconveniencing me! Rebased.

**Comment by mvdbeek** [comment]
Tests are clearly failing due to wheels issues.

File: lib/galaxy/webapps/galaxy/config_schema.yml [hunk 4]
```diff
@@ -766,6 +766,30 @@ mapping:
...
+        type: str
+        default: file
+        required: false
+        desc: |
+          Mulled resolution caching. Mulled resolution uses external APIs of quay.io, these
+          requests are caching using this and the following parameters
+
+      mulled_resolution_cache_data_dir:
+        type: str
+        default: citations/data
```
Comment (nsoranzo) [inline lib/galaxy/webapps/galaxy/config_schema.yml:779]: This seem a copy/paste mistake.

File: lib/galaxy/webapps/galaxy/config_schema.yml [hunk 5]
```diff
@@ -766,6 +766,30 @@ mapping:
...
+        path_resolves_to: data_dir
+        required: false
+        desc: |
+          Data directory used by beaker for caching mulled resolution requests.
+
+      mulled_resolution_cache_lock_dir:
+        type: str
+        default: citations/locks
```
Comment (nsoranzo) [inline lib/galaxy/webapps/galaxy/config_schema.yml:787]: Here too.

**Review by nsoranzo — COMMENTED** [review]
@jmchilton Can you please fix the 2 defaults and update config files with `make config-rebuild` ?
